Add support for 32KB multi_tensor_apply kernel arguments

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134373

## Benchmark

On H100 SXM (HBM2e, 500W TDP), CUDA Toolkit=12.2, Driver Version=535.154.05, with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa) (`torch._foreach_copy_`):

**Baseline**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp0g_x4sys
device ms: 0.891, cpu ms: 7.200
memory bandwidth: 1457.727 GB/s
```

Single iteration trace:
<img width="1432" alt="image" src="https://github.com/user-attachments/assets/8ef54365-0265-4281-a0f0-d4c2f448300e">


**This PR**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp3jqiugli
device ms: 0.683, cpu ms: 6.745
memory bandwidth: 1902.010 GB/s
```

Single iteration trace:
<img width="1074" alt="image" src="https://github.com/user-attachments/assets/e52acad1-d09b-492c-9611-6d69e339f3ac">

## Binary Size and Kernel Specialization
The binary size for `libtorch_cuda.so` increased 6MB (243MB -> 249MB).

```
// NOTE: [32KB kernel argument size support]
// 32KB kernel argument size support has three requirements:
// - CUDART_VERSION >= 12010
// - Driver version >= 530
// - GPU arch >= VOLTA
//
// Due to minor version compatibility, it possible for binaries built with
// CUDART_VERSION >= 12010 to run with driver version < 530. Since driver
// version can only be checked at runtime, if CUDART_VERSION >= 12010, we have
// to build both 4KB and 32KB kernels and determine the appropriate kernel to
// dispatch at runtime.
//
// - If CUDART_VERSION < 12010, only 4KB kernels will be instantiated.
//
// - If CUDART_VERSION >= 12010:
//   - Host code:
//     - We always instantiate the launching stub for both 4KB and 32KB kernels.
//   - Device code:
//     - If __CUDA_ARCH__ >= 700, we always instantiate both 4KB and 32KB
//     kernels.
//     - If __CUDA_ARCH__ < 700, it's not possible to even compile an empty
//     32KB kernel (formal parameter space overflowed). Thus, we only
//     instantiate a declaration for 32KB kernels. This is valid as long as the
//     declaration-only kernel is not launched.
//
// - At runtime, we dispatch to the 32KB kernel if driver version >= 530 and
// GPU arch >= VOLTA.
//
// - TODO(yifu): once there's a CUDART version that is not compatible with any
// driver version below 530, we can determine at compile time to not compile
// the kernels for 4KB kernel argument size.
//
// https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/
```

Commit:
[WIP] add support for 32KB multi_tensor_apply kernel arguments

[ghstack-poisoned]

Commit:
Update on "[WIP] add support for 32KB multi_tensor_apply kernel arguments"




[ghstack-poisoned]

Commit:
Update on "[WIP] add support for 32KB multi_tensor_apply kernel arguments"




[ghstack-poisoned]

eqy commented:
Looks like bazel build is failing for some reason?

crcrpar commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -23,6 +23,84 @@ static constexpr int depth_to_max_tensors_scalarlist_of_complex_double[2] = {
     72,
     60};
 
+// NOTE: [32KB kernel argument size support]
+// 32KB kernel argument size support is available only when CUDART_VERSION >=
+// 12010 and the driver version is >= 530. Only the former condition can be
+// checked at compile time. This implies:
+//
+// - If CUDART_VERSION < 12010, kernels for the 32KB kernel argument size will
+// not be built.
+//
+// - If CUDART_VERSION >= 12010, kernels for both 4KB and 32KB kernel argument
+// sizes will be built. However, due to CUDA's minor version compatibility,
+// even when CUDART_VERSION >= 12010, the driver may not support 32KB.
+// Therefore, a runtime check is necessary to determine the appropriate kernel
+// to dispatch.
+//
+// - TODO(yifu): once there's a CUDART version that is not compatible with any
+// driver version below 530, we can determine at compile time to not compile
+// the kernels for 4KB kernel argument size.
```
nice comment

crcrpar commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -38,60 +116,96 @@ __device__ __forceinline__ void load_store(
-                       [depth_to_max_tensors_scalarlist_of_complex_double[0]];
-  int64_t
-      numel_for_tensor[depth_to_max_tensors_scalarlist_of_complex_double[0]];
+template <bool large_kernel_arg>
```
wouldn't the comment one line above need update?

crcrpar reviewed (approved):
great update, thank you for working on this :)

Commit:
Update on "Add support for 32KB multi_tensor_apply kernel arguments"


## Benchmark

On H100 SXM (HBM2e, 500W TDP), CUDA Toolkit=12.2, Driver Version=535.154.05, with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa) (`torch._foreach_copy_`):

**Baseline**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp0g_x4sys
device ms: 0.891, cpu ms: 7.200
memory bandwidth: 1457.727 GB/s
```

Single iteration trace:
<img width="1432" alt="image" src="https://github.com/user-attachments/assets/8ef54365-0265-4281-a0f0-d4c2f448300e">


**This PR**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp3jqiugli
device ms: 0.683, cpu ms: 6.745
memory bandwidth: 1902.010 GB/s
```

Single iteration trace:
<img width="1074" alt="image" src="https://github.com/user-attachments/assets/e52acad1-d09b-492c-9611-6d69e339f3ac">

## Binary Size and Kernel Specialization
The binary size for `libtorch_cuda.so` increased 6MB (243MB -> 249MB).

```
// NOTE: [32KB kernel argument size support]
// 32KB kernel argument size support is available only when CUDART_VERSION >=
// 12010 and the driver version is >= 530. Only the former condition can be
// checked at compile time. This implies:
//
// - If CUDART_VERSION < 12010, kernels for the 32KB kernel argument size will
// not be built.
//
// - If CUDART_VERSION >= 12010, kernels for both 4KB and 32KB kernel argument
// sizes will be built. However, due to CUDA's minor version compatibility,
// even when CUDART_VERSION >= 12010, the driver may not support 32KB.
// Therefore, a runtime check is necessary to determine the appropriate kernel
// to dispatch.
//
// - TODO(yifu): once there's a CUDART version that is not compatible with any
// driver version below 530, we can determine at compile time to not compile
// the kernels for 4KB kernel argument size.
```

[ghstack-poisoned]

Commit:
Update on "Add support for 32KB multi_tensor_apply kernel arguments"


## Benchmark

On H100 SXM (HBM2e, 500W TDP), CUDA Toolkit=12.2, Driver Version=535.154.05, with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa) (`torch._foreach_copy_`):

**Baseline**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp0g_x4sys
device ms: 0.891, cpu ms: 7.200
memory bandwidth: 1457.727 GB/s
```

Single iteration trace:
<img width="1432" alt="image" src="https://github.com/user-attachments/assets/8ef54365-0265-4281-a0f0-d4c2f448300e">


**This PR**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp3jqiugli
device ms: 0.683, cpu ms: 6.745
memory bandwidth: 1902.010 GB/s
```

Single iteration trace:
<img width="1074" alt="image" src="https://github.com/user-attachments/assets/e52acad1-d09b-492c-9611-6d69e339f3ac">

## Binary Size and Kernel Specialization
The binary size for `libtorch_cuda.so` increased 6MB (243MB -> 249MB).

```
// NOTE: [32KB kernel argument size support]
// 32KB kernel argument size support is available only when CUDART_VERSION >=
// 12010 and the driver version is >= 530. Only the former condition can be
// checked at compile time. This implies:
//
// - If CUDART_VERSION < 12010, kernels for the 32KB kernel argument size will
// not be built.
//
// - If CUDART_VERSION >= 12010, kernels for both 4KB and 32KB kernel argument
// sizes will be built. However, due to CUDA's minor version compatibility,
// even when CUDART_VERSION >= 12010, the driver may not support 32KB.
// Therefore, a runtime check is necessary to determine the appropriate kernel
// to dispatch.
//
// - TODO(yifu): once there's a CUDART version that is not compatible with any
// driver version below 530, we can determine at compile time to not compile
// the kernels for 4KB kernel argument size.
```

[ghstack-poisoned]

Commit:
Update on "Add support for 32KB multi_tensor_apply kernel arguments"


## Benchmark

On H100 SXM (HBM2e, 500W TDP), CUDA Toolkit=12.2, Driver Version=535.154.05, with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa) (`torch._foreach_copy_`):

**Baseline**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp0g_x4sys
device ms: 0.891, cpu ms: 7.200
memory bandwidth: 1457.727 GB/s
```

Single iteration trace:
<img width="1432" alt="image" src="https://github.com/user-attachments/assets/8ef54365-0265-4281-a0f0-d4c2f448300e">


**This PR**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp3jqiugli
device ms: 0.683, cpu ms: 6.745
memory bandwidth: 1902.010 GB/s
```

Single iteration trace:
<img width="1074" alt="image" src="https://github.com/user-attachments/assets/e52acad1-d09b-492c-9611-6d69e339f3ac">

## Binary Size and Kernel Specialization
The binary size for `libtorch_cuda.so` increased 6MB (243MB -> 249MB).

```
// NOTE: [32KB kernel argument size support]
// 32KB kernel argument size support is available only when CUDART_VERSION >=
// 12010 and the driver version is >= 530. Only the former condition can be
// checked at compile time. This implies:
//
// - If CUDART_VERSION < 12010, kernels for the 32KB kernel argument size will
// not be built.
//
// - If CUDART_VERSION >= 12010, kernels for both 4KB and 32KB kernel argument
// sizes will be built. However, due to CUDA's minor version compatibility,
// even when CUDART_VERSION >= 12010, the driver may not support 32KB.
// Therefore, a runtime check is necessary to determine the appropriate kernel
// to dispatch.
//
// - TODO(yifu): once there's a CUDART version that is not compatible with any
// driver version below 530, we can determine at compile time to not compile
// the kernels for 4KB kernel argument size.
```

[ghstack-poisoned]

Commit:
Update on "Add support for 32KB multi_tensor_apply kernel arguments"


## Benchmark

On H100 SXM (HBM2e, 500W TDP), CUDA Toolkit=12.2, Driver Version=535.154.05, with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa) (`torch._foreach_copy_`):

**Baseline**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp0g_x4sys
device ms: 0.891, cpu ms: 7.200
memory bandwidth: 1457.727 GB/s
```

Single iteration trace:
<img width="1432" alt="image" src="https://github.com/user-attachments/assets/8ef54365-0265-4281-a0f0-d4c2f448300e">


**This PR**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp3jqiugli
device ms: 0.683, cpu ms: 6.745
memory bandwidth: 1902.010 GB/s
```

Single iteration trace:
<img width="1074" alt="image" src="https://github.com/user-attachments/assets/e52acad1-d09b-492c-9611-6d69e339f3ac">

## Binary Size and Kernel Specialization
The binary size for `libtorch_cuda.so` increased 6MB (243MB -> 249MB).

```
// NOTE: [32KB kernel argument size support]
// 32KB kernel argument size support is available only when CUDART_VERSION >=
// 12010 and the driver version is >= 530. Only the former condition can be
// checked at compile time. This implies:
//
// - If CUDART_VERSION < 12010, kernels for the 32KB kernel argument size will
// not be built.
//
// - If CUDART_VERSION >= 12010, kernels for both 4KB and 32KB kernel argument
// sizes will be built. However, due to CUDA's minor version compatibility,
// even when CUDART_VERSION >= 12010, the driver may not support 32KB.
// Therefore, a runtime check is necessary to determine the appropriate kernel
// to dispatch.
//
// - TODO(yifu): once there's a CUDART version that is not compatible with any
// driver version below 530, we can determine at compile time to not compile
// the kernels for 4KB kernel argument size.
```

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -14,14 +14,93 @@ static constexpr int64_t kILP = 4;
+    int driver_ver = 0;
+    cudaDriverGetVersion(&driver_ver);
+    cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+    *supports_large_kernel_arg_ = (driver_ver >= 12010) && prop->major >= 7;
```
Driver version 530 has numerical value 12010 in this struct?

albanD reviewed (commented):
Nice!

FYI @seemethere for the non-trivial binary size increase. I think it is worth it given the speedup and we can in theory recover it in the future when nvidia stops supporting driver 530.

eqy commented:
Out of curiosity, what was the fix for the bazel build issue, was it actually build related or because that particular job was running on an unsupported GPU?

janeyx99 commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -14,14 +14,93 @@ static constexpr int64_t kILP = 4;
+// https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/
+bool supports_large_kernel_arg() {
+#if !defined(USE_ROCM) && defined(CUDART_VERSION) && CUDART_VERSION >= 12010
+  static std::optional<bool> supports_large_kernel_arg_ = std::nullopt;
+  if (!supports_large_kernel_arg_.has_value()) {
```
looks like a noob q: when would line 39 ever be False if line 38 sets it to null? Is `supports_large_kernel_arg_1 accessible from outside this function?

janeyx99 replied:
OOOOH i missed the static. Very cool. It has the same effect as function caching then.

yifuwang commented:
> was it actually build related or because that particular job was running on an unsupported GPU

@eqy it's both. The bazel build was using an M40 GPU and the feature actually required [Volta and above](https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/). We have to handle __CUDA_ARCH__ < 700 specially at compile time, which wasn't done in the previous version. I'll include a comment to describe this.

janeyx99 reviewed (commented):
Thanks for pursuing this change!

Commit:
Update on "Add support for 32KB multi_tensor_apply kernel arguments"


## Benchmark

On H100 SXM (HBM2e, 500W TDP), CUDA Toolkit=12.2, Driver Version=535.154.05, with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa) (`torch._foreach_copy_`):

**Baseline**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp0g_x4sys
device ms: 0.891, cpu ms: 7.200
memory bandwidth: 1457.727 GB/s
```

Single iteration trace:
<img width="1432" alt="image" src="https://github.com/user-attachments/assets/8ef54365-0265-4281-a0f0-d4c2f448300e">


**This PR**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp3jqiugli
device ms: 0.683, cpu ms: 6.745
memory bandwidth: 1902.010 GB/s
```

Single iteration trace:
<img width="1074" alt="image" src="https://github.com/user-attachments/assets/e52acad1-d09b-492c-9611-6d69e339f3ac">

## Binary Size and Kernel Specialization
The binary size for `libtorch_cuda.so` increased 6MB (243MB -> 249MB).

```
// NOTE: [32KB kernel argument size support]
// 32KB kernel argument size support is available only when CUDART_VERSION >=
// 12010 and the driver version is >= 530. Only the former condition can be
// checked at compile time. This implies:
//
// - If CUDART_VERSION < 12010, kernels for the 32KB kernel argument size will
// not be built.
//
// - If CUDART_VERSION >= 12010, kernels for both 4KB and 32KB kernel argument
// sizes will be built. However, due to CUDA's minor version compatibility,
// even when CUDART_VERSION >= 12010, the driver may not support 32KB.
// Therefore, a runtime check is necessary to determine the appropriate kernel
// to dispatch.
//
// - TODO(yifu): once there's a CUDART version that is not compatible with any
// driver version below 530, we can determine at compile time to not compile
// the kernels for 4KB kernel argument size.
```

[ghstack-poisoned]

janeyx99 commented:
@yifuwang I think this PR is good to go

yifuwang commented:
@pytorchbot merge

Commit:
Update on "Add support for 32KB multi_tensor_apply kernel arguments"


## Benchmark

On H100 SXM (HBM2e, 500W TDP), CUDA Toolkit=12.2, Driver Version=535.154.05, with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa) (`torch._foreach_copy_`):

**Baseline**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp0g_x4sys
device ms: 0.891, cpu ms: 7.200
memory bandwidth: 1457.727 GB/s
```

Single iteration trace:
<img width="1432" alt="image" src="https://github.com/user-attachments/assets/8ef54365-0265-4281-a0f0-d4c2f448300e">


**This PR**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp3jqiugli
device ms: 0.683, cpu ms: 6.745
memory bandwidth: 1902.010 GB/s
```

Single iteration trace:
<img width="1074" alt="image" src="https://github.com/user-attachments/assets/e52acad1-d09b-492c-9611-6d69e339f3ac">

## Binary Size and Kernel Specialization
The binary size for `libtorch_cuda.so` increased 6MB (243MB -> 249MB).

```
// NOTE: [32KB kernel argument size support]
// 32KB kernel argument size support has three requirements:
// - CUDART_VERSION >= 12010
// - Driver version >= 530
// - GPU arch >= VOLTA
//
// Due to minor version compatibility, it possible for binaries built with
// CUDART_VERSION >= 12010 to run with driver version < 530. Since driver
// version can only be checked at runtime, if CUDART_VERSION >= 12010, we have
// to build both 4KB and 32KB kernels and determine the appropriate kernel to
// dispatch at runtime.
//
// - If CUDART_VERSION < 12010, only 4KB kernels will be instantiated.
//
// - If CUDART_VERSION >= 12010:
//   - Host code:
//     - We always instantiate the launching stub for both 4KB and 32KB kernels.
//   - Device code:
//     - If __CUDA_ARCH__ >= 700, we always instantiate both 4KB and 32KB
//     kernels.
//     - If __CUDA_ARCH__ < 700, it's not possible to even compile an empty
//     32KB kernel (formal parameter space overflowed). Thus, we only
//     instantiate a declaration for 32KB kernels. This is valid as long as the
//     declaration-only kernel is not launched.
//
// - At runtime, we dispatch to the 32KB kernel if driver version >= 530 and
// GPU arch >= VOLTA.
//
// - TODO(yifu): once there's a CUDART version that is not compatible with any
// driver version below 530, we can determine at compile time to not compile
// the kernels for 4KB kernel argument size.
//
// https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/
```

[ghstack-poisoned]

yifuwang commented:
The previous merge attempt failed due to an MSVC build issue. Trying to merge again with a possible fix.
```
// MSVC has a problem with constexpr and can't handle passing them to templates
// as arguments. We need to replace it with const static.
// https://github.com/state-spaces/mamba/issues/12#issuecomment-1848835662
```

@pytorchbot merge

Commit:
Update on "Add support for 32KB multi_tensor_apply kernel arguments"


## Benchmark

On H100 SXM (HBM2e, 500W TDP), CUDA Toolkit=12.2, Driver Version=535.154.05, with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa) (`torch._foreach_copy_`):

**Baseline**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp0g_x4sys
device ms: 0.891, cpu ms: 7.200
memory bandwidth: 1457.727 GB/s
```

Single iteration trace:
<img width="1432" alt="image" src="https://github.com/user-attachments/assets/8ef54365-0265-4281-a0f0-d4c2f448300e">


**This PR**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp3jqiugli
device ms: 0.683, cpu ms: 6.745
memory bandwidth: 1902.010 GB/s
```

Single iteration trace:
<img width="1074" alt="image" src="https://github.com/user-attachments/assets/e52acad1-d09b-492c-9611-6d69e339f3ac">

## Binary Size and Kernel Specialization
The binary size for `libtorch_cuda.so` increased 6MB (243MB -> 249MB).

```
// NOTE: [32KB kernel argument size support]
// 32KB kernel argument size support has three requirements:
// - CUDART_VERSION >= 12010
// - Driver version >= 530
// - GPU arch >= VOLTA
//
// Due to minor version compatibility, it possible for binaries built with
// CUDART_VERSION >= 12010 to run with driver version < 530. Since driver
// version can only be checked at runtime, if CUDART_VERSION >= 12010, we have
// to build both 4KB and 32KB kernels and determine the appropriate kernel to
// dispatch at runtime.
//
// - If CUDART_VERSION < 12010, only 4KB kernels will be instantiated.
//
// - If CUDART_VERSION >= 12010:
//   - Host code:
//     - We always instantiate the launching stub for both 4KB and 32KB kernels.
//   - Device code:
//     - If __CUDA_ARCH__ >= 700, we always instantiate both 4KB and 32KB
//     kernels.
//     - If __CUDA_ARCH__ < 700, it's not possible to even compile an empty
//     32KB kernel (formal parameter space overflowed). Thus, we only
//     instantiate a declaration for 32KB kernels. This is valid as long as the
//     declaration-only kernel is not launched.
//
// - At runtime, we dispatch to the 32KB kernel if driver version >= 530 and
// GPU arch >= VOLTA.
//
// - TODO(yifu): once there's a CUDART version that is not compatible with any
// driver version below 530, we can determine at compile time to not compile
// the kernels for 4KB kernel argument size.
//
// https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/
```

[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge -r

yifuwang commented:
@pytorchbot merge

drisspg commented:
@pytorchbot revert "See https://github.com/pytorch/pytorch/issues/135126 for more details"

drisspg commented:
@pytorchbot revert  -m "See https://github.com/pytorch/pytorch/issues/135126 for more details"

drisspg commented:
@pytorchbot revert -m "See https://github.com/pytorch/pytorch/issues/135126 for more details" -c weird

malfet commented:
I think one needs to add driver version check before using the feature..

yifuwang commented:
@mafet we already do a driver version check in this PR - when building with eligible toolkit for eligbile CUDA_ARCH, we build the 32KB kernels in addition to 4KB kernels. At runtime, we only choose the 32KB kernels if the driver version is eligible. Otherwise, we fallback to the 4KB kernels.

It's having a very weird issue when running with minor version compatibility (e.g. 525 + 12.3). I actually verifed the PR with this combination of driver + toolkit version and did not run into issues. However, digging into https://github.com/pytorch/pytorch/issues/135126, I found that if we call `cudaGetDeviceCount` before we perform a device allocation, the device allocation would fail with "operation not supported" error. If we perform a device allocation first, the failure goes away. **The failure didn't even involve `multi_tensor_apply`, so I suspect that the driver is unhappy about what's in nvfatbin and it only happen in certain scenarios.**

This is unfortunate. If this PR's approach doesn't work, it means that we can only enable this feature when the user tells us the driver version the binary would run on at compile time. It also means that we can not enable this feature in our pre-built binaries (unless we build another version for 530+) until the next major toolkit version.
